File: LeetCode/src/unionfind/NumberOfIslands.java
package unionfind;

public class NumberOfIslands {
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        UnionFind uf = new UnionFind(grid);

        int[][] directions = {{0, 1}, {1, 0}};

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    int curIndex = i * cols + j;
                    for (int[] dir : directions) {
                        int row = i + dir[0];
                        int col = j + dir[1];
                        if (row < rows && col < cols && grid[row][col] == '1') {
                            int neiIndex = row * cols + col;
                            uf.union(curIndex, neiIndex);
                        }
                    }
                }
            }
        }

        return uf.getCount();
    }

    class UnionFind {
        private int[] parent;
        private int[] rank;
        private int count;

        public UnionFind(char[][] grid) {
            int rows = grid.length;
            int cols = grid[0].length;

            parent = new int[rows * cols];
            rank = new int[rows * cols];

            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < cols; j++) {
                    if (grid[i][j] == '1') {
                        int index = i * cols + j;
                        parent[index] = index;
                        rank[index] = 0;
                        count++;
                    }
                }
            }
        }

        public int find(int p) {
            if (parent[p] != p) {
                parent[p] = find(parent[p]);
            }
            return parent[p];
        }

        public void union(int p, int q) {
            int rootP = find(p);
            int rootQ = find(q);

            if (rootP != rootQ) {
                if (rank[rootP] > rank[rootQ]) {
                    parent[rootQ] = rootP;
                } else if (rank[rootP] < rank[rootQ]) {
                    parent[rootP] = rootQ;
                } else {
                    parent[rootQ] = rootP;
                    rank[rootP]++;
                }
                count--;
            }
        }

        public int getCount() {
            return count;
        }
    }

    public static void main(String[] args) {
        NumberOfIslands solution = new NumberOfIslands();

        char[][] grid = {
                {'1', '1', '0', '0', '0'},
                {'1', '1', '0', '0', '0'},
                {'0', '0', '1', '0', '0'},
                {'0', '0', '0', '1', '1'}
        };

        System.out.println("Number of islands: " + solution.numIslands(grid));// Output: 3
    }
}

File: LeetCode_Python/graph/unionfind/NumberOfIslands.py
class UnionFind:
    def __init__(self, grid: list[list[str]]):
        self.m = len(grid)
        self.n = len(grid[0]) if self.m else 0
        size = self.m * self.n
        self.parent = [-1] * size
        self.rank = [0] * size
        self.count = 0

        for i in range(self.m):
            for j in range(self.n):
                if grid[i][j] == "1":
                    index = i * self.n + j
                    self.parent[index] = index
                    self.count += 1
    
    def find(self, p: int) -> int:
        if self.parent[p] != p:
            self.parent[p] = self.find(self.parent[p])
        return self.parent[p]

    def union(self, p: int, q: int) -> None:
        root_p = self.find(p)
        root_q = self.find(q)
        if root_p == root_q:
            return
        
        if self.rank[root_p] > self.rank[root_q]:
            self.parent[root_q] = root_p
        elif self.rank[root_p] < self.rank[root_q]:
            self.parent[root_p] = root_q
        else:
            self.parent[root_q] = root_p
            self.rank[root_p] += 1
        self.count -= 1

class NumberOfIslands:
  def numIslands(self, grid: list[list[str]]) -> int:
      m, n = len(grid), len(grid[0])
      uf = UnionFind(grid)

      directions = [(0, 1), (1, 0)]
      for i in range(m):
        for j in range(n):
            if grid[i][j] == '1':
              for d in directions:
                  r, c = i + d[0], j + d[1]
                  if 0 <= r < m and 0 <= c < n and grid[r][c] == '1':
                      uf.union(i * n + j, r * n + c)
        
      return uf.count
  
if __name__ == "__main__":
    solution = NumberOfIslands()
    grid1 = [
        ["1","1","1","1","0"],
        ["1","1","0","1","0"],
        ["1","1","0","0","0"],
        ["0","0","0","0","0"]
    ]
    print("Expected Output: 1")
    print("Actual Output:", solution.numIslands([row[:] for row in grid1])) 
    
    grid2 = [
        ["1","1","0","0","0"],
        ["1","1","0","0","0"],
        ["0","0","1","0","0"],
        ["0","0","0","1","1"]
    ]

    print("Expected Output: 3")
    print("Actual Output:", solution.numIslands([row[:] for row in grid2])) 